### PR TITLE
Woo: Try "zipper" action-list state in chaining actions.

### PR DESCRIPTION
**Pull request by coderkevin**
_Note: This PR is incomplete by itself. It temporarily breaks some functionality in the product save code. It needs later PRs to be complete. After all PRs have been reviewed, I will merge them down into this one, test again, and merge it all to master at one time._

Reviews needed (checked when done)

- [x] Core refactor PR (this PR)
- [x] Readme update (#15484)
- [x] Additional test for carrying data state (#15468)
- [x] Product Create Implementation (#15483)

The existing code had one overarching problem, and it was that each action could not be executed dynamically, but must be created statically before starting the action-list. Furthermore, those actions themselves had to be serializable, so they couldn't contain functions themselves. This made it increasing difficult to update things like placeholder IDs as the list moved on. In addition to this, the way middleware dispatches leads to a first-in-last-out resolution of actions. This means that the actions don't resolve chronologically in the reducer, even though they were dispatched in chronological order.

How does this PR solve it?

1. It holds the action list in action properties and passes it on to the next action. This removes the requirement for the action list to be serializable.

2. It replaces static actions for each step with an `onStep` function that is passed the dispatch function and the current version of the action list. This allows for dynamic creation of actions and modification of the action list itself to add data for next steps, etc.

3. It organizes the step list into `nextSteps`, `currentStep` and `prevSteps`, so index numbers no longer have to be tracked.

To Test: `npm test`

### CODE REVIEW

File: client/extensions/woocommerce/state/data-layer/action-list/index.js [hunk 1]
```diff
@@ -8,76 +8,92 @@ const debug = debugFactor( 'woocommerce:action-list' );
...
+export default {
+	[ WOOCOMMERCE_ACTION_LIST_STEP_NEXT ]: [ handleStepNext ],
+	[ WOOCOMMERCE_ACTION_LIST_STEP_SUCCESS ]: [ handleStepSuccess ],
+	[ WOOCOMMERCE_ACTION_LIST_STEP_FAILURE ]: [ handleStepFailure ],
+};
+
+export function handleStepNext( { dispatch }, action, now = Date.now() ) {
+	const { actionList } = action;
+	const { prevSteps, currentStep, nextSteps } = actionList;
+	const [ nextStep, ...remainingSteps ] = nextSteps || [];
```
Comment (justinshreve) [inline client/extensions/woocommerce/state/data-layer/action-list/index.js]: Nice :).

**Comment by justinshreve** [comment]
Tip for reviewing: Review #15484 (the README) before or during this review.

File: client/extensions/woocommerce/state/action-list/test/fixtures/index.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+export const time = {
+	stepAStart: Date.now() - 3000,
+	stepAEnd: Date.now() - 3100,
+	stepBStart: Date.now() - 2000,
+	stepBEnd: Date.now() - 1100,
+	stepCStart: Date.now() - 1000,
+	stepCEnd: Date.now() - 500,
+	stepEStart: Date.now() - 2500,
+	stepEEnd: Date.now() - 2200,
```
Comment (dmsnell) [inline client/extensions/woocommerce/state/action-list/test/fixtures/index.js]: what out for timing issues with your tests. we have `sinon.useFakeTimers()` to provide controlled clocks

Comment (coderkevin) [inline client/extensions/woocommerce/state/action-list/test/fixtures/index.js]: For this test, the timestamps are inconsequential to the data flow and for UI/inspection purposes only. So they don't actually matter. But I'll definitely look into the fake timers for future work--especially where timing actually matters.

**Comment by dmsnell** [comment]
woo wee is this a confusing diff! big one too - may it work without fail for you

**Comment by coderkevin** [comment]
No worries, @dmsnell, this was the result of 4 other PRs that were separately reviewed merged into this one before merging into master. I did it this way so I could do a big refactor, get each part reviewed, then merge it all at once. That way, master is never in an "in-between" broken state.

**Pull request merged**